components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs

```diff
@@ -31,20 +31,97 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
+                var difference = value - _valueChangesCount;
+                _valueChangesCount = value;
+                ValueChangesCountLastMinute = difference;
+            }
+        }
+
+        /// <inheritdoc/>
+        public ulong DataChangesCountLastMinute {
+            get => CalculateSumForRingBuffer(_dataChangesBuffer, ref _lastPointerDataChanges, _bucketWidth, _lastWriteTimeDataChange);
+            private set => IncreaseRingBuffer(_dataChangesBuffer, ref _lastPointerDataChanges, _bucketWidth, value, ref _lastWriteTimeDataChange);
+        }
+
+        /// <inheritdoc/>
+        public ulong DataChangesCount {
+            get {
+                return _dataChangesCount;
+            }
+            private set {
+                var difference = value - _dataChangesCount;
+                _dataChangesCount = value;
+                DataChangesCountLastMinute = difference;
+            }
+        }
+
+        /// <summary>
+        /// Iterates the array and add up all values
+        /// </summary>
+        private static ulong CalculateSumForRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, DateTime lastWriteTime) {
+
+            // if IncreaseRingBuffer wasn't called for some time, maybe some stale values are included
+            var now = DateTime.UtcNow;
+            var indexPointer = now.Second % bucketWidth;
+            CleanStaleRingBufferBuckets(array, ref lastPointer, bucketWidth, indexPointer, lastWriteTime, now);
+
+            // with cleaned buffer, we can just accumulate all buckets
+            ulong sum = 0;
+            for(int index = 0; index< array.Length; index++) {
+                sum += array[index];
+            }
+            return sum;
+        }
+
+        /// <summary>
+        /// Helper function to distribute values over array based on time
+        /// </summary>
+        private static void IncreaseRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, ulong difference, ref DateTime lastWriteTime) {
+            var now = DateTime.UtcNow;
+            var indexPointer = now.Second % bucketWidth;
+            CleanStaleRingBufferBuckets(array, ref lastPointer, bucketWidth, indexPointer, lastWriteTime, now);
+
+            array[indexPointer] += difference;
+            lastPointer = indexPointer;
+            lastWriteTime = now;
+        }
+
+        /// <summary>
+        /// Empty the ring buffer buckets if necessary
+        /// </summary>
+        private static void CleanStaleRingBufferBuckets(ulong[] array, ref int lastPointer, int bucketWidth, int indexPointer, DateTime lastWrite, DateTime now) {
+
+            // if last update was > bucketsize seconds in the past delete whole array
+            if (lastWrite != DateTime.MinValue) {
+                var deleteWholeArray = (now - lastWrite).TotalSeconds > bucketWidth;
+                if (deleteWholeArray) {
+                    Array.Clear(array, 0, array.Length);
+                    return;
```

Comment:
We need to update `lastPointer` in this case as well. I believe that it should be assigned the value of `indexPointer`.

Reply:
no it is updated in line 94, when the write happens

Reply:
changed :ok:

Reply:
anyway changed for better readability